File: backend-dotnet-layer/Services/WorkingMemoryStore.cs
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using Microsoft.Extensions.Options;
using Microsoft.SemanticKernel;
using Microsoft.SemanticKernel.ChatCompletion;
using BackendDotnetLayer.Configuration;

namespace BackendDotnetLayer.Services;

public sealed class WorkingMemoryStore
{
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web);

    private readonly HttpClient _httpClient;
    private readonly ILogger<WorkingMemoryStore> _logger;
    private readonly WorkingMemoryOptions _options;

    public WorkingMemoryStore(
        HttpClient httpClient,
        IOptions<WorkingMemoryOptions> options,
        ILogger<WorkingMemoryStore> logger)
    {
        _httpClient = httpClient;
        _options = options.Value;
        _logger = logger;
    }

    public async Task<IReadOnlyList<ChatMessageContent>> GetMessagesAsync(string sessionId, CancellationToken cancellationToken)
    {
        var baseUrl = GetAgentMemoryServerUrl();
        if (string.IsNullOrWhiteSpace(baseUrl))
        {
            return [];
        }

        var requestUri = $"{baseUrl}/v1/working-memory/{Uri.EscapeDataString(sessionId)}?namespace={Uri.EscapeDataString(_options.Namespace)}";

        try
        {
            using var request = new HttpRequestMessage(HttpMethod.Get, requestUri);
            using var response = await _httpClient.SendAsync(request, cancellationToken);
            if (!response.IsSuccessStatusCode)
            {
                return [];
            }

            var payload = await response.Content.ReadAsStringAsync(cancellationToken);
            var root = JsonNode.Parse(payload);
            var messages = root?["messages"]?.AsArray();
            if (messages is null)
            {
                return [];
            }

            var result = new List<ChatMessageContent>();
            foreach (var messageNode in messages)
            {
                var role = messageNode?["role"]?.GetValue<string>() ?? string.Empty;
                var content = messageNode?["content"]?.GetValue<string>() ?? string.Empty;
                var message = ToChatMessage(role, content);
                if (message is not null)
                {
                    result.Add(message);
                }
            }

            return result;
        }
        catch (Exception exception)
        {
            _logger.LogError(exception, "Error loading working memory for session {SessionId}", sessionId);
            return [];
        }
    }

    public async Task UpdateMessagesAsync(string sessionId, IReadOnlyList<ChatMessageContent> messages, CancellationToken cancellationToken)
    {
        var baseUrl = GetAgentMemoryServerUrl();
        if (string.IsNullOrWhiteSpace(baseUrl))
        {
            return;
        }

        var messagesToStore = messages
            .Where(ShouldStore)
            .Select(message => new Dictionary<string, string>
            {
                ["role"] = DetermineRole(message),
                ["content"] = message.Content ?? string.Empty
            })
            .ToList();

        var requestBody = new
        {
            session_id = sessionId,
            messages = messagesToStore,
            @namespace = _options.Namespace,
            ttl_seconds = _options.TimeToLiveInSeconds,
            long_term_memory_strategy = new
            {
                strategy = "discrete",
                config = new { }
            }
        };

        try
        {
            using var request = new HttpRequestMessage(HttpMethod.Put, $"{baseUrl}/v1/working-memory/{Uri.EscapeDataString(sessionId)}")
            {
                Content = new StringContent(JsonSerializer.Serialize(requestBody, JsonOptions), Encoding.UTF8, "application/json")
            };

            await _httpClient.SendAsync(request, cancellationToken);
        }
        catch (Exception exception)
        {
            _logger.LogError(exception, "Error updating working memory for session {SessionId}", sessionId);
        }
    }

    public async Task DeleteMessagesAsync(string sessionId, CancellationToken cancellationToken)
    {
        var baseUrl = GetAgentMemoryServerUrl();
        if (string.IsNullOrWhiteSpace(baseUrl))
        {
            return;
        }

        var requestUri = $"{baseUrl}/v1/working-memory/{Uri.EscapeDataString(sessionId)}?namespace={Uri.EscapeDataString(_options.Namespace)}";

        try
        {
            using var request = new HttpRequestMessage(HttpMethod.Delete, requestUri);
            await _httpClient.SendAsync(request, cancellationToken);
        }
        catch (Exception exception)
        {
            _logger.LogError(exception, "Error deleting working memory for session {SessionId}", sessionId);
        }
    }

    private string? GetAgentMemoryServerUrl()
    {
        return string.IsNullOrWhiteSpace(_options.AgentMemoryServerUrl)
            ? Environment.GetEnvironmentVariable("AGENT_MEMORY_SERVER_URL")
            : _options.AgentMemoryServerUrl;
    }

    private bool ShouldStore(ChatMessageContent message)
    {
        if (message.Role == AuthorRole.System)
        {
            return _options.StoreSystemMessages;
        }

        if (message.Role == AuthorRole.Assistant)
        {
            return _options.StoreAssistantMessages;
        }

        if (message.Role.Label.Equals("tool", StringComparison.OrdinalIgnoreCase))
        {
            return _options.StoreToolMessages;
        }

        return true;
    }

    private static string DetermineRole(ChatMessageContent message)
    {
        if (message.Role == AuthorRole.User)
        {
            return "user";
        }

        if (message.Role == AuthorRole.Assistant)
        {
            return "assistant";
        }

        if (message.Role == AuthorRole.System)
        {
            return "system";
        }

        return message.Role.Label;
    }

    private ChatMessageContent? ToChatMessage(string role, string content)
    {
        if ((!_options.StoreSystemMessages && role.Equals("system", StringComparison.OrdinalIgnoreCase)) ||
            (!_options.StoreAssistantMessages && (role.Equals("assistant", StringComparison.OrdinalIgnoreCase) || role.Equals("ai", StringComparison.OrdinalIgnoreCase))) ||
            (!_options.StoreToolMessages && role.Equals("tool", StringComparison.OrdinalIgnoreCase)))
        {
            return null;
        }

        return role.ToLowerInvariant() switch
        {
            "user" => new ChatMessageContent(AuthorRole.User, content),
            "assistant" or "ai" => new ChatMessageContent(AuthorRole.Assistant, content),
            "system" => new ChatMessageContent(AuthorRole.System, content),
            "tool" => new ChatMessageContent(new AuthorRole("tool"), content),
            _ => null
        };
    }
}

File: backend-layer/src/main/java/io/redis/devrel/workshop/extensions/WorkingMemoryStore.java
package io.redis.devrel.workshop.extensions;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import dev.langchain4j.data.message.*;
import dev.langchain4j.store.memory.chat.ChatMemoryStore;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.*;
import java.util.stream.Collectors;

@Component
public class WorkingMemoryStore implements ChatMemoryStore {

    private static final Logger logger = LoggerFactory.getLogger(WorkingMemoryStore.class);

    @Autowired
    private ObjectMapper objectMapper;

    @Autowired
    private HttpClient httpClient;

    private final String agentMemoryServerUrl;
    private long timeToLiveInSeconds = 300;
    private boolean storeSystemMessages = false;
    private boolean storeAiMessages = false;
    private boolean storeToolMessages = false;
    private String namespace = "short-term-memory";

    public WorkingMemoryStore(String agentMemoryServerUrl) {
        this.agentMemoryServerUrl = agentMemoryServerUrl;
    }

    @Override
    public List<ChatMessage> getMessages(Object memoryId) {
        List<ChatMessage> chatMessages = new ArrayList<>();
        var request = HttpRequest.newBuilder()
                .uri(URI.create(agentMemoryServerUrl + "/v1/working-memory/" +
                        memoryId + "?namespace=" + namespace))
                .GET()
                .build();

        try {
            var response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
            if (response.statusCode() == HttpStatus.OK.value()) {
                var messages = objectMapper.readTree(response.body()).path("messages");
                if (!messages.isEmpty() && !messages.isMissingNode()) {
                    for (JsonNode messageNode : messages) {
                        var role = messageNode.path("role").asText("");
                        var content = messageNode.path("content").asText("");

                        // Skip messages based on configuration
                        if ((!storeSystemMessages && "system".equalsIgnoreCase(role)) ||
                                (!storeAiMessages && "ai".equalsIgnoreCase(role)) ||
                                (!storeToolMessages && "tool".equalsIgnoreCase(role))) {
                            continue;
                        }

                        ChatMessage chatMessage = switch (role.toLowerCase()) {
                            case "user" -> UserMessage.from(content);
                            case "assistant", "ai" -> AiMessage.from(content);
                            case "system" -> SystemMessage.from(content);
                            case "tool" -> null;
                            default -> {
                                if (!role.isEmpty()) {
                                    logger.warn("Unknown message role: {}", role);
                                }
                                yield null;
                            }
                        };

                        if (chatMessage != null) {
                            chatMessages.add(chatMessage);
                        }
                    }

                    return chatMessages;
                }
            }
        } catch (Exception ex) {
            logger.error("Error during working-term memory search", ex);
        }

        return chatMessages;
    }

    @Override
    public void updateMessages(Object memoryId, List<ChatMessage> list) {
        try {
            // Filter out system and AI messages based on configuration
            List<ChatMessage> messagesToStore = list.stream()
                    .filter(msg -> storeSystemMessages || !(msg instanceof SystemMessage))
                    .filter(msg -> storeAiMessages || !(msg instanceof AiMessage))
                    .filter(msg -> storeToolMessages || !(msg instanceof ToolExecutionResultMessage))
                    .toList();

            List<Map<String, String>> messages = messagesToStore.stream()
                    .map(message -> {
                        Map<String, String> messageMap = new HashMap<>();
                        messageMap.put("role", determineRole(message));
                        messageMap.put("content", messageContent(message));
                        return messageMap;
                    })
                    .collect(Collectors.toList());

            Map<String, Object> requestBody = new HashMap<>();
            requestBody.put("session_id", memoryId.toString());
            requestBody.put("messages", messages);
            requestBody.put("namespace", namespace);
            requestBody.put("ttl_seconds", timeToLiveInSeconds);
            requestBody.put("long_term_memory_strategy",
                    Map.of("strategy", "discrete",
                            "config", Map.of()));

            String jsonPayload = objectMapper.writeValueAsString(requestBody);

            var request = HttpRequest.newBuilder()
                    .uri(URI.create(agentMemoryServerUrl + "/v1/working-memory/" + memoryId))
                    .header("Content-Type", "application/json")
                    .PUT(HttpRequest.BodyPublishers.ofString(jsonPayload))
                    .build();

            httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        } catch (Exception ex) {
            logger.error("Error updating working memory for session: {}", memoryId, ex);
        }
    }

    @Override
    public void deleteMessages(Object memoryId) {
        try {
            var request = HttpRequest.newBuilder()
                    .uri(URI.create(agentMemoryServerUrl + "/v1/working-memory/" +
                            memoryId + "?namespace=" + namespace))
                    .DELETE()
                    .build();

            var response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() == HttpStatus.OK.value() ||
                    response.statusCode() == HttpStatus.NO_CONTENT.value() ||
                    response.statusCode() == HttpStatus.ACCEPTED.value()) {
                logger.info("Successfully deleted chat messages for session: {}", memoryId);
            } else if (response.statusCode() == HttpStatus.NOT_FOUND.value()) {
                logger.warn("Chat messages not found for session: {}", memoryId);
            } else {
                logger.error("Failed to delete chat messages. Status code: {}, Response: {}",
                        response.statusCode(), response.body());
            }

        } catch (Exception ex) {
            logger.error("Error deleting chat messages for session: " + memoryId, ex);
        }
    }

    private String determineRole(ChatMessage message) {
        return switch (message) {
            case UserMessage userMessage -> "user";
            case AiMessage aiMessage -> "assistant";
            case SystemMessage systemMessage -> "system";
            case ToolExecutionResultMessage toolExecutionResultMessage -> "tool";
            default -> {
                String className = message.getClass().getSimpleName().toLowerCase();
                if (className.contains("user")) {
                    yield "user";
                } else if (className.contains("ai") || className.contains("assistant")) {
                    yield "assistant";
                } else if (className.contains("system")) {
                    yield "system";
                } else if (className.contains("tool")) {
                    yield "tool";
                } else {
                    logger.warn("Unknown ChatMessage type: {}", message.getClass().getName());
                    yield "unknown";
                }
            }
        };
    }

    private String messageContent(ChatMessage message) {
        return switch (message) {
            case UserMessage userMessage -> userMessage.singleText();
            case AiMessage aiMessage -> aiMessage.text();
            case SystemMessage systemMessage -> systemMessage.text();
            case ToolExecutionResultMessage toolExecutionResultMessage -> toolExecutionResultMessage.text();
            default -> {
                logger.warn("Unknown message type for content extraction: {}", message.getClass().getName());
                yield message.toString();
            }
        };
    }

    public long getTimeToLiveInSeconds() {
        return timeToLiveInSeconds;
    }

    public void setTimeToLiveInSeconds(long timeToLiveInSeconds) {
        this.timeToLiveInSeconds = timeToLiveInSeconds;
    }

    public boolean isStoreSystemMessages() {
        return storeSystemMessages;
    }

    public void setStoreSystemMessages(boolean storeSystemMessages) {
        this.storeSystemMessages = storeSystemMessages;
    }

    public boolean isStoreAiMessages() {
        return storeAiMessages;
    }

    public void setStoreAiMessages(boolean storeAiMessages) {
        this.storeAiMessages = storeAiMessages;
    }

    public boolean isStoreToolMessages() {
        return storeToolMessages;
    }

    public void setStoreToolMessages(boolean storeToolMessages) {
        this.storeToolMessages = storeToolMessages;
    }

    public String getNamespace() {
        return namespace;
    }

    public void setNamespace(String namespace) {
        this.namespace = namespace;
    }

    public static Builder builder() {
        return new Builder();
    }

    public static class Builder {
        private String agentMemoryServerUrl;
        private Optional<Long> timeToLiveInSeconds = Optional.empty();
        private Optional<Boolean> storeSystemMessages = Optional.empty();
        private Optional<Boolean> storeAiMessages = Optional.empty();
        private Optional<Boolean> storeToolMessages = Optional.empty();
        private Optional<String> namespace = Optional.empty();

        public Builder agentMemoryServerUrl(String value) {
            this.agentMemoryServerUrl = value;
            return this;
        }

        public Builder timeToLiveInSeconds(long value) {
            this.timeToLiveInSeconds = Optional.of(value);
            return this;
        }

        public Builder storeSystemMessages(boolean value) {
            this.storeSystemMessages = Optional.of(value);
            return this;
        }

        public Builder storeAiMessages(boolean value) {
            this.storeAiMessages = Optional.of(value);
            return this;
        }

        public Builder storeToolMessages(boolean value) {
            this.storeToolMessages = Optional.of(value);
            return this;
        }

        public Builder namespace(String value) {
            this.namespace = Optional.of(value);
            return this;
        }

        public WorkingMemoryStore build() {
            if (agentMemoryServerUrl == null) {
                throw new IllegalStateException("agentMemoryServerUrl is required");
            }

            WorkingMemoryStore workingMemoryStore = new WorkingMemoryStore(agentMemoryServerUrl);
            timeToLiveInSeconds.ifPresent(workingMemoryStore::setTimeToLiveInSeconds);
            storeSystemMessages.ifPresent(workingMemoryStore::setStoreSystemMessages);
            storeAiMessages.ifPresent(workingMemoryStore::setStoreAiMessages);
            storeToolMessages.ifPresent(workingMemoryStore::setStoreToolMessages);
            namespace.ifPresent(workingMemoryStore::setNamespace);

            return workingMemoryStore;
        }
    }
}
